step-framework-timeseries/src/main/java/step/core/timeseries/ingestion/TimeSeriesIngestionPipeline.java

```diff
@@ -0,0 +1,186 @@
+        if (logger.isTraceEnabled()) {
+            logger.trace("Ingesting point. Attributes=" + attributes.toString() + ", Timestamp=" + timestamp + ", Value=" + value);
+        }
+        lock.readLock().lock();
+        try {
+            long index = timestampToBucketTimestamp(timestamp, sourceResolution);
+            Map<Map<String, Object>, BucketBuilder> bucketsForTimestamp = seriesQueue.computeIfAbsent(index, k -> new ConcurrentHashMap<>());
+            bucketsForTimestamp.computeIfAbsent(attributes, k -> BucketBuilder.create(index).withAttributes(new BucketAttributes(attributes))).ingest(value);
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    // TODO flush method can be improved with a batcher (run when a specific timeout ends, or a specific amount of data is collected)
+    public void flush() {
+        flush(true);
+    }
+
+    private void flush(boolean flushAll) {
+        lock.writeLock().lock();
+        try {
+            trace("Flushing");
+            long now = System.currentTimeMillis();
+
+            seriesQueue.forEach((k, v) -> {
+                if (flushAll || k < now - FLUSH_OFFSET) {
+                    // Remove the entry from the map and iterate over it afterwards
+                    // This enables concurrent execution of flushing and ingestion
+                    seriesQueue.remove(k).forEach((attributes, bucketBuilder) -> {
+                        Bucket bucket = bucketBuilder.build();
+                        saveOrMerge(bucket);
+                        if (flushCallback != null) {
+                            flushCallback.accept(bucket);
+                        }
+                    });
+                }
+            });
+
+            flushCount.increment();
+            trace("Flushed");
+        } catch (Throwable e) {
+            logger.error("Error while flushing", e);
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    private void saveOrMerge(Bucket newBucket) {
+        if (mergeBucketsOnFlush) {
+            Equals beginFilter = Filters.equals("begin", newBucket.getBegin());
+            Filter attributesFilter = getAttributesFilter(newBucket.getAttributes());
+            Optional<Bucket> foundBucket = collection.find(Filters.and(Arrays.asList(beginFilter, attributesFilter)), null, null, null, 0).findAny();
+            if (foundBucket.isPresent()) {
+                Bucket existingBucket = foundBucket.get();
+                Bucket accumulatedBucket = new BucketBuilder(existingBucket.getBegin())
+                        .withAttributes(existingBucket.getAttributes())
+                        .accumulate(existingBucket)
+                        .accumulate(newBucket)
+                        .build();
+                accumulatedBucket.setId(existingBucket.getId());
+                collection.save(accumulatedBucket);
+                return;
+            }
+        }
+        collection.save(newBucket);
```

Comment:
This seems wrong, you're saving the accumulatedBucket and here again the new Bucket

Reply:
@david-stephan that's why I added the return on line 143, or do I miss something?

Reply:
@lucian-baciu I missed it, for code readability, this is not the best, but acceptable.

Reply:
I replaced the return with another save